src/bulk/common.ts

```diff
@@ -1,43 +1,198 @@
-import { PromiseProvider } from '../promise_provider';
-import { Long, ObjectId } from '../bson';
-import { MongoError, MongoWriteConcernError } from '../error';
 import {
   applyWriteConcern,
   applyRetryableWrites,
   executeLegacyOperation,
   isPromiseLike,
   hasAtomicOperators,
   maxWireVersion,
-  Callback
+  Callback,
+  MongoDBNamespace
 } from '../utils';
+import { PromiseProvider } from '../promise_provider';
+import { Long, ObjectId, Document, BSONSerializeOptions } from '../bson';
+import { MongoError, MongoWriteConcernError } from '../error';
 import { executeOperation } from '../operations/execute_operation';
 import { InsertOperation } from '../operations/insert';
 import { UpdateOperation } from '../operations/update';
 import { DeleteOperation } from '../operations/delete';
-import type { WriteConcern } from '../write_concern';
+import { WriteConcern, WriteConcernOptions } from '../write_concern';
+import type { HasRetryableWrites } from './../utils';
+import type { Collection } from '../collection';
+import type { Topology } from '../sdam/topology';
+import type { Hint } from './../operations/operation';
+import type { addToOperationsList as orderedAddToOperationsList } from './ordered';
+import type { addToOperationsList as unorderedAddToOperationsList } from './unordered';
+import type { CollationOptions } from '../cmap/wire_protocol/write_command';
+import type { InsertOptions } from './../operations/insert';
+import type { RemoveOptions } from './../cmap/wire_protocol/index';
+import type { ReplaceOptions } from './../operations/replace_one';
+import type { UpdateOptions } from './../operations/update';
+
+interface FinalOptionsConfigOptions {
+  bypassDocumentValidation?: boolean;
+  retryWrites?: boolean;
+}
+interface ResultHandler {
+  (err?: any, result?: any): void;
+  operationId?: number;
+}
 
 // Error codes
 const WRITE_CONCERN_ERROR = 64;
 
 // Insert types
-const INSERT = 1;
-const UPDATE = 2;
-const REMOVE = 3;
+export const INSERT = 1;
+export const UPDATE = 2;
+export const REMOVE = 3;
+
+/** Number indicating the document type */
+export type BatchTypes = typeof INSERT | typeof UPDATE | typeof REMOVE;
+
+/** @public */
+export interface Operation {
```

Comment:
[naming nit] It's not named the same as OperationBase so that's not an issue but it might be confusing that its generically named but scoped to dealing with bulk api stuff, BulkOperation though could be confused with BulkOperationBase? but maybe that's a preferential confusion 😆

Reply:
Yes agree on this one, too generic, I'm not sure if this thing is already defined and can just be `AnyModel`? Not sure. Down for suggestions.

Reply:
`BulkOperationBase` 👍